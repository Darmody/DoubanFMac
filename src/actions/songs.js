// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const listen = (channel: number): Action => ({
  type: types.SONG_LISTEN_REQUEST,
  payload: channel,
})

export const next = (channel: number, sid?: number): Action => ({
  type: types.SONG_NEXT_REQUEST,
  payload: {
    channel,
    sid,
  }
})

export const mark = (channel: number, sid: number): Action => ({
  type: types.SONG_MARK_REQUEST,
  payload: {
    sid,
    channel,
  }
})

export const current = (song: {}): Action => ({
  type: types.SONG_CURRENT,
  payload: song,
})
