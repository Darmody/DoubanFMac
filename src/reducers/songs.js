// @flow
import Immutable from 'seamless-immutable'
import * as types from 'constants/types/ActionTypes'
import type { Reducer } from 'constants/types/Redux'

const initialState = Immutable({
  id: undefined,
  playedIds: [],
})

const reducer: Reducer = (state = initialState, { type, payload = {} }) => {
  switch (type) {
    case types.SONG_CURRENT: return state.merge({
      id: payload.sid,
    })
    case types.SONG_PLAYED_SUCCESS: return state.merge({
      playedIds: payload,
    })
    default: return state
  }
}

export default reducer
