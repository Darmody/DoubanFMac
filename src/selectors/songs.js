// @flow
import { SONG } from 'schemas'
import type { StateShape } from 'constants/types/Redux'
import { selectEntity } from 'selectors/entities'

export const selectCurrent = (state: StateShape) => selectEntity(
  state, SONG, state.songs.id,
)
