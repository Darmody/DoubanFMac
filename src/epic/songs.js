// @flow
import { Observable as Rx$ } from 'rxjs'
import * as types from 'constants/types/ActionTypes'
import type { Epic } from 'constants/types/Redux'
import * as API from 'clients/doubanRxClient'
import { current } from 'actions/songs'
import { update as updateEntities } from 'actions/entities'
import { normalizeResponse, getToken, fullfilled } from 'utils/operators'
import { SONG } from 'schemas'

const fetchSong = (type, api) => (action$, store) => action$
  .ofType(type)
  .pluck('payload')
  .switchMap(api(getToken(store)))
  .pluck('response', 'song')
  .mergeMap(([song]) => Rx$.merge(
    Rx$.of(normalizeResponse(song, SONG)),
    Rx$.of(current(song)),
  ))

const nextEpic: Epic = fetchSong(types.SONG_NEXT_REQUEST, API.nextSong)
const listenEpic: Epic = fetchSong(types.SONG_LISTEN_REQUEST, API.listenSong)
const banEpic: Epic = fetchSong(types.SONG_BAN_REQUEST, API.banSong)

const markEpic: Epic = (action$, store) => action$
  .ofType(types.SONG_MARK_REQUEST)
  .pluck('payload')
  .mergeMap(API.markSong(getToken(store)))
  .map(fullfilled(types.SONG_MARK_SUCCESS))

const likeEpic: Epic = (action$, store) => {
  const likeFullfilled = sid => () => Rx$.of(updateEntities([{
    field: ['songs', sid, 'like'],
    value: 1,
  }]))

  return action$
    .ofType(types.SONG_LIKE_REQUEST)
    .pluck('payload')
    .mergeMap(payload => API
      .likeSong(getToken(store))(payload)
      .mergeMap(likeFullfilled(payload.sid))
    )
}

const dislikeEpic: Epic = (action$, store) => {
  const dislikeFullfilled = sid => () => Rx$.of(updateEntities([{
    field: ['songs', sid, 'like'],
    value: 0,
  }]))

  return action$
    .ofType(types.SONG_DISLIKE_REQUEST)
    .pluck('payload')
    .mergeMap(payload => API
      .dislikeSong(getToken(store))(payload)
      .mergeMap(dislikeFullfilled(payload.sid))
    )
}

export default [
  banEpic,
  dislikeEpic,
  likeEpic,
  listenEpic,
  markEpic,
  nextEpic,
]
