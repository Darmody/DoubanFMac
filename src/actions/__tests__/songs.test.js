import * as actions from '../songs'

describe('action::songs', () => {
  it('创建一个 ban action', () => {
    expect(actions.ban(1, 2)).toMatchSnapshot()
  })

  it('创建一个 current action', () => {
    expect(actions.current({ sid: 1 })).toMatchSnapshot()
  })

  it('创建一个 like action', () => {
    expect(actions.like(1, 2)).toMatchSnapshot()
  })

  it('创建一个 dislike action', () => {
    expect(actions.dislike(1, 2)).toMatchSnapshot()
  })

  it('创建一个 listen action', () => {
    expect(actions.listen(1)).toMatchSnapshot()
  })

  it('创建一个 next action', () => {
    expect(actions.next(1, 2)).toMatchSnapshot()
  })

  it('创建一个 mark action', () => {
    expect(actions.mark(1, 2)).toMatchSnapshot()
  })

  it('创建一个 playedList action', () => {
    expect(actions.playedList(0, 10)).toMatchSnapshot()
  })
})
