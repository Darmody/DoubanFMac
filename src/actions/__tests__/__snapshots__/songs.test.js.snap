// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action::songs 创建一个 ban action 1`] = `
Object {
  "payload": Object {
    "channel": 1,
    "sid": 2,
  },
  "type": "SONGS/BAN/REQUEST",
}
`;

exports[`action::songs 创建一个 current action 1`] = `
Object {
  "payload": Object {
    "sid": 1,
  },
  "type": "SONGS/CURRENT",
}
`;

exports[`action::songs 创建一个 dislike action 1`] = `
Object {
  "payload": Object {
    "channel": 1,
    "sid": 2,
  },
  "type": "SONGS/DISLIKE/REQUEST",
}
`;

exports[`action::songs 创建一个 like action 1`] = `
Object {
  "payload": Object {
    "channel": 1,
    "sid": 2,
  },
  "type": "SONGS/LIKE/REQUEST",
}
`;

exports[`action::songs 创建一个 listen action 1`] = `
Object {
  "payload": 1,
  "type": "SONGS/LISTEN/REQUEST",
}
`;

exports[`action::songs 创建一个 mark action 1`] = `
Object {
  "payload": Object {
    "channel": 1,
    "sid": 2,
  },
  "type": "SONGS/MARK/REQUEST",
}
`;

exports[`action::songs 创建一个 next action 1`] = `
Object {
  "payload": Object {
    "channel": 1,
    "sid": 2,
  },
  "type": "SONGS/NEXT/REQUEST",
}
`;

exports[`action::songs 创建一个 playedList action 1`] = `
Object {
  "payload": Object {
    "limit": 10,
    "start": 0,
  },
  "type": "SONGS/PLAYED/REQUEST",
}
`;
